File: qabelbox/src/main/java/de/qabel/qabelbox/adapter/QabelSyncAdapter.java
package de.qabel.qabelbox.adapter;

import android.accounts.Account;
import android.content.AbstractThreadedSyncAdapter;
import android.content.BroadcastReceiver;
import android.content.ContentProviderClient;
import android.content.ContentResolver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.SyncResult;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.util.Log;

import java.sql.SQLException;
import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

import javax.inject.Inject;

import de.qabel.core.config.Contact;
import de.qabel.core.config.Contacts;
import de.qabel.core.config.Identities;
import de.qabel.core.config.Identity;
import de.qabel.desktop.repository.ContactRepository;
import de.qabel.desktop.repository.IdentityRepository;
import de.qabel.desktop.repository.exception.EntityNotFoundExcepion;
import de.qabel.desktop.repository.exception.PersistenceException;
import de.qabel.desktop.repository.sqlite.AndroidClientDatabase;
import de.qabel.qabelbox.QabelBoxApplication;
import de.qabel.qabelbox.chat.ChatMessageInfo;
import de.qabel.qabelbox.chat.ChatMessageItem;
import de.qabel.qabelbox.chat.ChatNotificationManager;
import de.qabel.qabelbox.chat.ChatServer;
import de.qabel.qabelbox.chat.SyncAdapterChatNotificationManager;
import de.qabel.qabelbox.helper.Helper;
import de.qabel.qabelbox.persistence.RepositoryFactory;
import de.qabel.qabelbox.services.DropConnector;
import de.qabel.qabelbox.services.HttpDropConnector;

public class QabelSyncAdapter extends AbstractThreadedSyncAdapter {

    private static final String TAG = "QabelSyncAdapter";
    ContentResolver mContentResolver;
    @Inject Context context;
    @Inject IdentityRepository identityRepository;
    @Inject ContactRepository contactRepository;
    @Inject ChatNotificationManager notificationManager;
    @Inject ChatServer chatServer;
    DropConnector dropConnector;
    private List<ChatMessageInfo> currentMessages = new ArrayList<>();
    private BroadcastReceiver receiver;

    public QabelSyncAdapter(Context context, boolean autoInitialize) {
        super(context, autoInitialize);
        init(context);
    }

    @Inject
    public void setDropConnector(DropConnector dropConnector) {
        this.dropConnector = dropConnector;
    }

    public QabelSyncAdapter(
            Context context,
            boolean autoInitialize,
            boolean allowParallelSyncs) {
        super(context, autoInitialize, allowParallelSyncs);
        init(context);
    }

    private void init(Context context) {
        this.context = context;
        mContentResolver = context.getContentResolver();
        QabelBoxApplication.getApplicationComponent(context).inject(this);
        registerNotificationReceiver();
    }

    private void registerNotificationReceiver() {
        IntentFilter filter = new IntentFilter(Helper.INTENT_REFRESH_CONTACTLIST);
        filter.setPriority(0);
        receiver = new BroadcastReceiver() {
            @Override
            public void onReceive(Context context, Intent intent) {
                notificationManager.updateNotifications(currentMessages);
            }
        };
        context.registerReceiver(receiver, filter);
    }



    @Override
    public void onPerformSync(
            Account account,
		    Bundle extras,
		    String authority,
		    ContentProviderClient provider,
		    SyncResult syncResult) {
        Log.w(TAG, "Starting drop message sync");
        Set<Identity> identities;
        try {
            identities = identityRepository.findAll().getIdentities();
        } catch (PersistenceException e) {
            Log.e(TAG, "Sync failed", e);
            return;
        }
        List<ChatMessageItem> retrievedMessages = new ArrayList<>();
        for (Identity identity: identities) {
            Log.i(TAG, "Loading messages for identity "+ identity.getAlias());
            retrievedMessages.addAll(
                    chatServer.refreshList(dropConnector, identity));
        }
        notifyForNewMessages(retrievedMessages);
    }

    void notifyForNewMessages(List<ChatMessageItem> retrievedMessages) {
        if (retrievedMessages.size() == 0) {
            return;
        }
        updateNotificationManager(retrievedMessages);
        Intent notificationIntent = new Intent(Helper.INTENT_SHOW_NOTIFICATION);
        context.sendOrderedBroadcast(notificationIntent, null);
        Intent refresh = new Intent(Helper.INTENT_REFRESH_CONTACTLIST);
        context.sendBroadcast(refresh);
    }

    private void updateNotificationManager(List<ChatMessageItem> retrievedMessages) {
        currentMessages = toChatMessageInfo(retrievedMessages);
    }

    private List<ChatMessageInfo> toChatMessageInfo(List<ChatMessageItem> retrievedMessages) {
        List<ChatMessageInfo> messages = new ArrayList<>();
        for (ChatMessageItem msg: retrievedMessages) {
            try {
                Identity identity = identityRepository.find(msg.getReceiverKey());
                Contact contact = contactRepository.findByKeyId(identity, msg.getSenderKey());
                ChatMessageInfo messageInfo = new ChatMessageInfo(
                        contact,
                        identity,
                        msg.getData().getMessage(),
                        new Date(msg.getTime()),
                        ChatMessageInfo.MessageType.MESSAGE);
                messages.add(messageInfo);
            } catch (EntityNotFoundExcepion | PersistenceException entityNotFoundExcepion) {
                Log.w(TAG, "Could not find contact " + msg.getSenderKey()
                        +" for identity " + msg.getReceiverKey());
            }
        }
        return messages;
    }

}

File: qabelbox/src/main/java/de/qabel/qabelbox/adapter/QabelSyncAdapter.kt
package de.qabel.qabelbox.adapter

import android.accounts.Account
import android.content.*
import android.os.Bundle
import android.util.Log
import de.qabel.core.config.Identity
import de.qabel.desktop.repository.ContactRepository
import de.qabel.desktop.repository.IdentityRepository
import de.qabel.desktop.repository.exception.EntityNotFoundExcepion
import de.qabel.desktop.repository.exception.PersistenceException
import de.qabel.qabelbox.QabelBoxApplication
import de.qabel.qabelbox.chat.ChatMessageInfo
import de.qabel.qabelbox.chat.ChatMessageItem
import de.qabel.qabelbox.chat.ChatNotificationManager
import de.qabel.qabelbox.chat.ChatServer
import de.qabel.qabelbox.helper.Helper
import de.qabel.qabelbox.services.DropConnector
import java.util.*
import javax.inject.Inject

open class QabelSyncAdapter : AbstractThreadedSyncAdapter {
    lateinit internal var mContentResolver: ContentResolver
    @Inject lateinit internal var context: Context
    @Inject lateinit internal var identityRepository: IdentityRepository
    @Inject lateinit internal var contactRepository: ContactRepository
    @Inject lateinit internal var notificationManager: ChatNotificationManager
    @Inject lateinit internal var chatServer: ChatServer
    @Inject lateinit internal var dropConnector: DropConnector
    private var currentMessages: List<ChatMessageInfo> = ArrayList()

    constructor(context: Context, autoInitialize: Boolean) : super(context, autoInitialize) {
        init(context)
    }

    @Inject
    fun setDropConnector(dropConnector: DropConnector) {
        this.dropConnector = dropConnector
    }

    constructor(
            context: Context,
            autoInitialize: Boolean,
            allowParallelSyncs: Boolean) : super(context, autoInitialize, allowParallelSyncs) {
        init(context)
    }

    private fun init(context: Context) {
        this.context = context
        mContentResolver = context.contentResolver
        QabelBoxApplication.getApplicationComponent(context).inject(this)
        registerNotificationReceiver()
    }

    private fun registerNotificationReceiver() {
        val filter = IntentFilter(Helper.INTENT_SHOW_NOTIFICATION)
        filter.priority = 0
        val receiver = object : BroadcastReceiver() {
            override fun onReceive(context: Context, intent: Intent) {
                notificationManager.updateNotifications(currentMessages)
            }
        }
        context.registerReceiver(receiver, filter)
    }


    override fun onPerformSync(
            account: Account,
            extras: Bundle,
            authority: String,
            provider: ContentProviderClient,
            syncResult: SyncResult) {
        Log.w(TAG, "Starting drop message sync")
        val identities: Set<Identity>
        try {
            identities = identityRepository.findAll().identities
        } catch (e: PersistenceException) {
            Log.e(TAG, "Sync failed", e)
            return
        }

        val retrievedMessages = ArrayList<ChatMessageItem>()
        for (identity in identities) {
            Log.i(TAG, "Loading messages for identity " + identity.alias)
            retrievedMessages.addAll(
                    chatServer.refreshList(dropConnector, identity))
        }
        notifyForNewMessages(retrievedMessages)
    }

    open fun notifyForNewMessages(retrievedMessages: List<ChatMessageItem>) {
        if (retrievedMessages.size == 0) {
            return
        }
        updateNotificationManager(retrievedMessages)
        val notificationIntent = Intent(Helper.INTENT_SHOW_NOTIFICATION)
        context.sendOrderedBroadcast(notificationIntent, null)
        val refreshList = Intent(Helper.INTENT_REFRESH_CONTACTLIST)
        context.sendBroadcast(refreshList)
        val refreshChat = Intent(Helper.INTENT_REFRESH_CHAT)
        context.sendBroadcast(refreshChat)
    }

    private fun updateNotificationManager(retrievedMessages: List<ChatMessageItem>) {
        currentMessages = toChatMessageInfo(retrievedMessages)
    }

    private fun toChatMessageInfo(retrievedMessages: List<ChatMessageItem>): List<ChatMessageInfo> {
        val messages = ArrayList<ChatMessageInfo>()
        for (msg in retrievedMessages) {
            try {
                val identity = identityRepository.find(msg.receiverKey)
                val contact = contactRepository.findByKeyId(identity, msg.senderKey)
                val messageInfo = ChatMessageInfo(
                        contact,
                        identity,
                        msg.data.message,
                        Date(msg.time),
                        ChatMessageInfo.MessageType.MESSAGE)
                messages.add(messageInfo)
            } catch (entityNotFoundExcepion: EntityNotFoundExcepion) {
                Log.w(TAG, "Could not find contact " + msg.senderKey
                        + " for identity " + msg.receiverKey)
            } catch (entityNotFoundExcepion: PersistenceException) {
                Log.w(TAG, "Could not find contact " + msg.senderKey + " for identity " + msg.receiverKey)
            }

        }
        return messages
    }

    companion object {
        private val TAG = "QabelSyncAdapter"
    }

}
